Convert OkHttpClient to Kotlin

ZacSweers opened the pull request:
Resolves #4747 

Renamed the file first with no changes to try to protect git history (unless you want to squash)

ZacSweers commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -0,0 +1,831 @@
+     * @deprecated [SSLSocketFactory] does not expose its [X509TrustManager], which is a field that
+     * OkHttp needs to build a clean certificate chain. This method instead must use reflection to
+     * extract the trust manager. Applications should prefer to call
+     * `sslSocketFactory(SSLSocketFactory, X509TrustManager)`, which avoids such reflection.
```
Can't link directly because https://github.com/Kotlin/dokka/issues/80

ZacSweers commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -0,0 +1,831 @@
+      isTLS = isTLS || spec.isTls
+    }
+
+    // TODO - these nullability warnings existed in Java
```
Let me know how you want to handle these

![image](https://user-images.githubusercontent.com/1361086/54588942-90f5ae00-49e1-11e9-8f4e-7828ab1c23c7.png)

swankjesse commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -13,881 +13,666 @@
-        builder.addLenient(line);
-      }
+  @get:JvmName("proxy")
+  val proxy: Proxy? = builder.proxy
```
By making it a public val rather than a private field, you’re changing the API for Kotlin callers. I don’t think we’re ready for this. In particular I’d like to defer changing APIs for Kotlin callers until we have an automatic migration path based on strong deprecation.

Could you change this PR to make zero API changes for Java and Kotlin callers? Then later in follow-up we can make things more Kotlin friendly, and support those changes by test cases.

ZacSweers replied:
Can do

PromanSEW commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -0,0 +1,829 @@
+import java.net.Socket
+import java.security.GeneralSecurityException
+import java.time.Duration
+import java.util.ArrayList
```
`ArrayList` has alias in Kotlin, so import is not needed

ZacSweers replied:
Good call! fa4cd4f

swankjesse commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -0,0 +1,829 @@
+ *
+ * ## Customize your client with newBuilder()
+ *
+ * You can customize a shared OkHttpClient instance with [.newBuilder]. This builds a
```
nit: I think IntelliJ prefers `[newBuilder]` over `[.newBuilder]`.

ZacSweers replied:
Yeah this was an auto-migrate thing. Thought I squashed all these, fixed with a few others in 1716088

swankjesse commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -0,0 +1,829 @@
+      isTLS = isTLS || spec.isTls
+    }
+
+    // TODO - these nullability warnings existed in Java
```
what’s actionable here? My preference is to just omit the TODO

ZacSweers replied:
I think it's a legit issue.

Either sslSocketFactory should be nullable here https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/OkHttpClient.java#L327

Because this line could possibly assign a nullable value to it even if the builder value is `null` if `isTLS` is `false`.

```kotlin
if (builder.sslSocketFactory != null || !isTLS) {
  this.sslSocketFactory = builder.sslSocketFactory
```

We can leave it with just the force-unwraps here, but the null-check below it suggested to me what it was something intended to be nullable. What do you think?

```kotlin
if (sslSocketFactory != null) {
  Platform.get().configureSslSocketFactory(sslSocketFactory)
}
```

And in either case, `builder.certificateChainCleaner` is implicitly assumed to be present and subsequently force-unwrapped.

swankjesse replied:
I think it's too strict as-is. Previously a null value would cause a theoretical NPE after calling the accessor; now you'll always get an NPE immediately when you construct.

Fix? Probably to make accessing this value throw if there's no TLS.

ZacSweers replied:
> Probably to make accessing this value throw if there's no TLS.

I think I didn't quite follow what `this` means in this context?

swankjesse replied:
I'd like to make the accessors return non-null values. If you attempt to read either on a client that doesn't have SSL, it should throw a UnsupportedOperationException.

swankjesse commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -0,0 +1,829 @@
+     * The default value is 10 seconds.
+     */
+    @IgnoreJRERequirement
+    fun connectTimeout(duration: Duration) = apply {
```
apply is a nice trick

yschimke replied:
That is great for the builder pattern, nice find!

swankjesse commented:
```
* What went wrong:
Execution failed for task ':okhttp:japicmp'.
> A failure occurred while executing Comparing [okhttp-4.0.0-SNAPSHOT.jar] with [okhttp-3.14.0.jar]
   > Detected binary changes between okhttp-4.0.0-SNAPSHOT.jar and okhttp-3.14.0.jar. See failure report at file:///home/travis/build/square/okhttp/okhttp/build/reports/japi.txt
```

JakeWharton commented:
We should dump that file to stdout when the task fails

On Sat, Mar 23, 2019, 6:13 PM Jesse Wilson <notifications@github.com wrote:

> * What went wrong:
> Execution failed for task ':okhttp:japicmp'.
> > A failure occurred while executing Comparing [okhttp-4.0.0-SNAPSHOT.jar] with [okhttp-3.14.0.jar]
>    > Detected binary changes between okhttp-4.0.0-SNAPSHOT.jar and okhttp-3.14.0.jar. See failure report at file:///home/travis/build/square/okhttp/okhttp/build/reports/japi.txt
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/square/okhttp/pull/4748#issuecomment-475909161>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEEEXxGH2YEwd-uPxuNRq-eC9wELaHnks5vZqbvgaJpZM4b7fE8>
> .
>

ZacSweers commented:
Here's a dump of the output. I missed that OkHttpClient wasn't `final` before, but the builder is final and I'm not sure if this is reporting a false positive?

Also the package private is expected. Right now it's changed to `internal`, but not sure what can be done about that.

lmk what you want to do!

```
Comparing binary compatibility of  against 
***! MODIFIED CLASS: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient  (not serializable)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.Authenticator authenticator()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.Cache cache()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) int callTimeoutMillis()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.CertificatePinner certificatePinner()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.ConnectionPool connectionPool()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.util.List connectionSpecs()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) int connectTimeoutMillis()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.CookieJar cookieJar()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.Dispatcher dispatcher()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.Dns dns()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.EventListener$Factory eventListenerFactory()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) boolean followRedirects()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) boolean followSslRedirects()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) javax.net.ssl.HostnameVerifier hostnameVerifier()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.util.List interceptors()
***! MODIFIED METHOD: PUBLIC (<- PACKAGE_PROTECTED) FINAL (<- NON_FINAL) okhttp3.internal.cache.InternalCache internalCache()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.util.List networkInterceptors()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder newBuilder()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) int pingIntervalMillis()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.util.List protocols()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.net.Proxy proxy()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.Authenticator proxyAuthenticator()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.net.ProxySelector proxySelector()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) int readTimeoutMillis()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) boolean retryOnConnectionFailure()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) javax.net.SocketFactory socketFactory()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) javax.net.ssl.SSLSocketFactory sslSocketFactory()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) int writeTimeoutMillis()
***! MODIFIED CLASS: PUBLIC STATIC FINAL okhttp3.OkHttpClient$Builder  (not serializable)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder addInterceptor(okhttp3.Interceptor)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder addNetworkInterceptor(okhttp3.Interceptor)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder authenticator(okhttp3.Authenticator)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient build()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder cache(okhttp3.Cache)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder callTimeout(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder callTimeout(java.time.Duration)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder certificatePinner(okhttp3.CertificatePinner)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder connectionPool(okhttp3.ConnectionPool)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder connectionSpecs(java.util.List)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder connectTimeout(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder connectTimeout(java.time.Duration)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder cookieJar(okhttp3.CookieJar)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder dispatcher(okhttp3.Dispatcher)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder dns(okhttp3.Dns)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder eventListener(okhttp3.EventListener)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder eventListenerFactory(okhttp3.EventListener$Factory)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder followRedirects(boolean)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder followSslRedirects(boolean)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder hostnameVerifier(javax.net.ssl.HostnameVerifier)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.util.List interceptors()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.util.List networkInterceptors()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder pingInterval(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder pingInterval(java.time.Duration)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder protocols(java.util.List)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder proxy(java.net.Proxy)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder proxyAuthenticator(okhttp3.Authenticator)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder proxySelector(java.net.ProxySelector)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder readTimeout(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder readTimeout(java.time.Duration)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder retryOnConnectionFailure(boolean)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder socketFactory(javax.net.SocketFactory)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder sslSocketFactory(javax.net.ssl.SSLSocketFactory)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder sslSocketFactory(javax.net.ssl.SSLSocketFactory, javax.net.ssl.X509TrustManager)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder writeTimeout(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder writeTimeout(java.time.Duration)
```

ZacSweers commented:
Output after making the client `open` again

```
Comparing binary compatibility of  against 
***! MODIFIED CLASS: PUBLIC okhttp3.OkHttpClient  (not serializable)
***! MODIFIED METHOD: PUBLIC (<- PACKAGE_PROTECTED) FINAL (<- NON_FINAL) okhttp3.internal.cache.InternalCache internalCache()
***! MODIFIED CLASS: PUBLIC STATIC FINAL okhttp3.OkHttpClient$Builder  (not serializable)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder addInterceptor(okhttp3.Interceptor)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder addNetworkInterceptor(okhttp3.Interceptor)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder authenticator(okhttp3.Authenticator)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient build()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder cache(okhttp3.Cache)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder callTimeout(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder callTimeout(java.time.Duration)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder certificatePinner(okhttp3.CertificatePinner)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder connectionPool(okhttp3.ConnectionPool)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder connectionSpecs(java.util.List)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder connectTimeout(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder connectTimeout(java.time.Duration)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder cookieJar(okhttp3.CookieJar)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder dispatcher(okhttp3.Dispatcher)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder dns(okhttp3.Dns)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder eventListener(okhttp3.EventListener)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder eventListenerFactory(okhttp3.EventListener$Factory)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder followRedirects(boolean)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder followSslRedirects(boolean)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder hostnameVerifier(javax.net.ssl.HostnameVerifier)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.util.List interceptors()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.util.List networkInterceptors()
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder pingInterval(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder pingInterval(java.time.Duration)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder protocols(java.util.List)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder proxy(java.net.Proxy)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder proxyAuthenticator(okhttp3.Authenticator)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder proxySelector(java.net.ProxySelector)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder readTimeout(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder readTimeout(java.time.Duration)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder retryOnConnectionFailure(boolean)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder socketFactory(javax.net.SocketFactory)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder sslSocketFactory(javax.net.ssl.SSLSocketFactory)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder sslSocketFactory(javax.net.ssl.SSLSocketFactory, javax.net.ssl.X509TrustManager)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder writeTimeout(long, java.util.concurrent.TimeUnit)
***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.OkHttpClient$Builder writeTimeout(java.time.Duration)

```

shaishavgandhi05 commented:
Some final methods are added to a `methodExcludes` like here. https://github.com/square/okhttp/blob/master/okhttp/build.gradle#L48

Will let @swankjesse confirm but here's a [PR](https://github.com/square/okhttp/pull/4771/files#diff-ea0d9417fdf1924df2fb03d1250c3a68R134) that added it to the excludes.

swankjesse commented:
@ZacSweers lemme know if you can fix & rebase, otherwise I’ll do it myself!

ZacSweers commented:
Rebased!

swankjesse commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -0,0 +1,820 @@
+        .withCertificateChainCleaner(certificateChainCleaner)
+  }
+
+  open fun dispatcher(): Dispatcher = dispatcher
```
These shouldn't be open. That's a bug in japicmp. Fix is to remove `open` everywhere and to suppress the warning in build.gradle.

We're waiting for japicmp to be released.

ZacSweers replied:
Is it a bug? It is technically not final in OkHttpClient.java: https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/OkHttpClient.java#L320, nor is the class. Happy to make them final here, but it would be a different change.

swankjesse replied:
OMG, you're right. I left it open for paralleluniverse.co.

I retract this request!

swankjesse commented on okhttp/src/main/java/okhttp3/OkHttpClient.kt:
```diff
@@ -0,0 +1,820 @@
+     * this client silently recovers from the following problems:
+     *
+     *  * **Unreachable IP addresses.** If the URL's host has multiple IP addresses,
+     * failure to reach any individual IP address doesn't fail the overall request. This can
```
Nit: indent the wrapped lines please.

ZacSweers replied:
452765d

swankjesse commented:
I’ve rebased, fixed tests, and am attempting to land this here:
https://github.com/square/okhttp/pull/4827
